[data] Fix HTTP streaming file download by using `open_input_stream`

xyuzh opened the pull request:
## What does this PR do?
   
   Fixes HTTP streaming file downloads in Ray Data's download operation. Some URIs (especially HTTP streams) require `open_input_stream` instead of `open_input_file`.
   
   ## Changes
   
   - Modified `download_bytes_threaded` in `plan_download_op.py` to try both `open_input_file` and `open_input_stream` for each URI
   - Improved error handling to distinguish between different error types
   - Failed downloads now return `None` gracefully instead of crashing
   
   ## Testing
```
import pyarrow as pa
from ray.data.context import DataContext
from ray.data._internal.planner.plan_download_op import download_bytes_threaded

# Test URLs: one valid, one 404
urls = [    
    "https://static-assets.tesla.com/configurator/compositor?context=design_studio_2?&bkba_opt=1&view=STUD_3QTR&size=600&model=my&options=$APBS,$IPB7,$PPSW,$SC04,$MDLY,$WY19P,$MTY46,$STY5S,$CPF0,$DRRH&crop=1150,647,390,180&",
]

# Create PyArrow table and call download function
table = pa.table({"url": urls})
ctx = DataContext.get_current()
results = list(download_bytes_threaded(table, ["url"], ["bytes"], ctx))

# Check results
result_table = results[0]
for i in range(result_table.num_rows):
    url = result_table['url'][i].as_py()
    bytes_data = result_table['bytes'][i].as_py()
    
    if bytes_data is None:
        print(f"Row {i}: FAILED (None) - try-catch worked ✓")
    else:
        print(f"Row {i}: SUCCESS ({len(bytes_data)} bytes)")
    print(f"  URL: {url[:60]}...")

print("\n✅ Test passed: Failed downloads return None instead of crashing.")
```

Before the fix:
```
TypeError: cannot set 'open_input_file' attribute of immutable type 'pyarrow._fs.FileSystem'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ray/default/test_streaming_fallback.py", line 110, in <module>
    test_download_expression_with_streaming_fallback()
  File "/home/ray/default/test_streaming_fallback.py", line 67, in test_download_expression_with_streaming_fallback
    with patch.object(pafs.FileSystem, "open_input_file", mock_open_input_file):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ray/anaconda3/lib/python3.12/unittest/mock.py", line 1594, in __enter__
    if not self.__exit__(*sys.exc_info()):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ray/anaconda3/lib/python3.12/unittest/mock.py", line 1603, in __exit__
    setattr(self.target, self.attribute, self.temp_original)
TypeError: cannot set 'open_input_file' attribute of immutable type 'pyarrow._fs.FileSystem'
(base) ray@ip-10-0-39-21:~/default$ python test.py
2025-11-11 18:32:23,510 WARNING util.py:1059 -- Caught exception in transforming worker!
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.12/site-packages/ray/data/_internal/util.py", line 1048, in _run_transforming_worker
    for result in fn(input_queue_iter):
                  ^^^^^^^^^^^^^^^^^^^^
  File "/home/ray/anaconda3/lib/python3.12/site-packages/ray/data/_internal/planner/plan_download_op.py", line 197, in load_uri_bytes
    yield f.read()
          ^^^^^^^^
  File "pyarrow/io.pxi", line 411, in pyarrow.lib.NativeFile.read
  File "pyarrow/io.pxi", line 263, in pyarrow.lib.NativeFile.size
  File "pyarrow/error.pxi", line 155, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 89, in pyarrow.lib.check_status
  File "/home/ray/anaconda3/lib/python3.12/site-packages/fsspec/implementations/http.py", line 743, in seek
    raise ValueError("Cannot seek streaming HTTP file")
ValueError: Cannot seek streaming HTTP file
Traceback (most recent call last):
  File "/home/ray/default/test.py", line 16, in <module>
    results = list(download_bytes_threaded(table, ["url"], ["bytes"], ctx))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ray/anaconda3/lib/python3.12/site-packages/ray/data/_internal/planner/plan_download_op.py", line 207, in download_bytes_threaded
    uri_bytes = list(
                ^^^^^
  File "/home/ray/anaconda3/lib/python3.12/site-packages/ray/data/_internal/util.py", line 1113, in make_async_gen
    raise item
  File "/home/ray/anaconda3/lib/python3.12/site-packages/ray/data/_internal/util.py", line 1048, in _run_transforming_worker
    for result in fn(input_queue_iter):
                  ^^^^^^^^^^^^^^^^^^^^
  File "/home/ray/anaconda3/lib/python3.12/site-packages/ray/data/_internal/planner/plan_download_op.py", line 197, in load_uri_bytes
    yield f.read()
          ^^^^^^^^
  File "pyarrow/io.pxi", line 411, in pyarrow.lib.NativeFile.read
  File "pyarrow/io.pxi", line 263, in pyarrow.lib.NativeFile.size
  File "pyarrow/error.pxi", line 155, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 89, in pyarrow.lib.check_status
  File "/home/ray/anaconda3/lib/python3.12/site-packages/fsspec/implementations/http.py", line 743, in seek
    raise ValueError("Cannot seek streaming HTTP file")
ValueError: Cannot seek streaming HTTP file
```
After the fix:
```
Row 0: SUCCESS (189370 bytes)
  URL: https://static-assets.tesla.com/configurator/compositor?cont...
```
   
   Tested with HTTP streaming URLs (e.g., Tesla configurator images) that previously failed:
   - ✅ Successfully downloads HTTP stream files
   - ✅ Gracefully handles failed downloads (returns None)
   - ✅ Maintains backward compatibility with existing file downloads

xyuzh committed:
[data] Fix HTTP streaming file download by using open_input_stream

Signed-off-by: xyuzh <xinyzng@gmail.com>

robertnishihara commented on python/ray/data/_internal/planner/plan_download_op.py:
```diff
@@ -189,12 +189,30 @@ def download_bytes_threaded(
+                            yield f.read()
+                        break
+                    except ValueError as e:
+                        # This handles the case where the file is not seekable and so we need to fall back to
```
We could make this more targeted and check that `e.args[0] == "Cannot seek streaming HTTP file"` but that could be a little brittle.

xyuzh replied:
Updated to use only `open_input_stream`

robertnishihara commented on python/ray/data/_internal/planner/plan_download_op.py:
```diff
@@ -189,12 +189,27 @@ def download_bytes_threaded(
+                        break  
+                    except ValueError as e:
+                        last_error = e
+                        continue
```
@xyuzh this comment might be correct. Please take a look. There might be some edge cases where we yield twice.

robertnishihara reviewed (commented):
Left some comments. Please also fix linting.

xyuzh commented on python/ray/data/_internal/planner/plan_download_op.py:
```diff
@@ -189,12 +189,27 @@ def download_bytes_threaded(
-                except OSError as e:
+                last_error = None
+                # Handle both file and stream for uri download
+                for uri_open_func in [fs.open_input_file, fs.open_input_stream]:
```
Updated to use only `open_input_stream`

xyuzh commented on python/ray/data/_internal/planner/plan_download_op.py:
```diff
@@ -189,12 +189,27 @@ def download_bytes_threaded(
+                        logger.error(f"Unexpected error in load_uri_bytes: {e}")
+                        yield None
+                        break
+                else:
```
Updated to use only `open_input_stream`

robertnishihara committed:
Update plan_download_op.py

Signed-off-by: Robert Nishihara <robertnishihara@gmail.com>

robertnishihara commented on python/ray/data/_internal/planner/plan_download_op.py:
```diff
@@ -190,13 +190,19 @@ def load_uri_bytes(uri_path_iterator):
+                except Exception as e:
+                    # Catch unexpected errors like pyarrow.lib.ArrowInvalid caused by an invalid uri like
+                    # `foo://bar` to avoid failing because of one invalid uri.
+                    logger.error(f"Unexpected error in load_uri_bytes: {e}")
```
@xyuzh we should also log the `uri_path` and `uri_column_name` in this message.

xyuzh committed:
[data] Improve error logging with column name and URI

Signed-off-by: xyuzh <xinyzng@gmail.com>

xyuzh committed:
[data] Use try-finally to ensure each URI yields exactly once

Prevents double-yield issue if file stream cleanup fails after successful read.

Signed-off-by: xyuzh <xinyzng@gmail.com>

xyuzh commented on python/ray/data/_internal/planner/plan_download_op.py:
```diff
@@ -190,13 +190,19 @@ def load_uri_bytes(uri_path_iterator):
+                    # Catch unexpected errors like pyarrow.lib.ArrowInvalid caused by an invalid uri like
+                    # `foo://bar` to avoid failing because of one invalid uri.
+                    logger.error(f"Unexpected error in load_uri_bytes: {e}")
+                    yield None
```
Now I added `finally` to resolve the yield twice possibility

xyuzh committed:
[data] Rename result to read_bytes for better clarity

Signed-off-by: xyuzh <xinyzng@gmail.com>

xyuzh committed:
[data] Change unexpected error logging from error to debug level

Since these errors are gracefully handled (return None), debug level is more appropriate.

Signed-off-by: xyuzh <xinyzng@gmail.com>

robertnishihara committed:
Update plan_download_op.py

Signed-off-by: Robert Nishihara <robertnishihara@gmail.com>

xyuzh committed:
[data] Add test for invalid URI scheme exception handling

Test that invalid URI schemes like foo://bar are caught and return None.

Signed-off-by: xyuzh <xinyzng@gmail.com>

xyuzh committed:
[data] Simplify invalid URI scheme test

Use ray.data.from_items and single test case for better clarity.

Signed-off-by: xyuzh <xinyzng@gmail.com>

xyuzh committed:
[data] Expand test to cover multiple malformed URI types

Test invalid scheme, missing scheme, empty URI, and relative path.

Signed-off-by: xyuzh <xinyzng@gmail.com>

xyuzh committed:
[data] Add more edge cases to malformed URI test

Include random strings, absolute/relative paths, injection attempts, null bytes, and line breaks.

Signed-off-by: xyuzh <xinyzng@gmail.com>

xyuzh committed:
[data] Simplify malformed URI test cases

Remove redundant path variations and consolidate to key edge cases.

Signed-off-by: xyuzh <xinyzng@gmail.com>

xyuzh committed:
[data] Consolidate invalid and malformed URI tests

Merge OSError test cases (file not found) into malformed URI test for better organization.

Signed-off-by: xyuzh <xinyzng@gmail.com>

xyuzh commented on python/ray/data/_internal/planner/plan_download_op.py:
@robertnishihara @bveeramani @richardliaw I found the `_resolve_paths_and_filesystem` raises exception for unresolved path and the exception is not handled here

To move forward I want to clarify:
* Are we supporting multiple columns of uri to be downloaded at the same time

The above matters to how we want to do exception handling when one uri fails.
`_resolve_paths_and_filesystem` considers the multiple uris in the same column to share the same file system and I don't think this is a good practice
As @robertnishihara also mentioned in [comment](https://github.com/ray-project/ray/pull/58464#issuecomment-3506941111)

bveeramani reviewed (commented):
@xyuzh Could you run the `read_from_uris_fixed_size` release test and ensure changing to `open_input_stream` doesn't introduce a performance regression? https://github.com/kyuds/ray/blob/0eb4ee2087b85497a58099568802fb59e5a8af7b/release/release_data_tests.yaml#L77-L78

robertnishihara committed:
Update test_download_expression.py

Signed-off-by: Robert Nishihara <robertnishihara@gmail.com>

robertnishihara committed:
Update test_download_expression.py

Signed-off-by: Robert Nishihara <robertnishihara@gmail.com>

robertnishihara committed:
Update test_download_expression.py

Signed-off-by: Robert Nishihara <robertnishihara@gmail.com>

robertnishihara committed:
Update test_download_expression.py

Signed-off-by: Robert Nishihara <robertnishihara@gmail.com>

bveeramani reviewed (approved):
LGTM other than elaborating on the comment.

If `open_input_stream` introduces a performance regression, we can fix-forward

robertnishihara committed:
Update test_download_expression.py

Signed-off-by: Robert Nishihara <robertnishihara@gmail.com>

robertnishihara commented on python/ray/data/tests/test_download_expression.py:
```diff
@@ -294,43 +294,33 @@ def test_download_expression_with_null_uris(self):
+            "local:///this/path/does/not/exist/file.txt",  # Invalid path
+            "",  # Empty URI
+            "foobar",  # Random string
+            # TODO(xyuzh): Add the tests below back once the issue is fixed.
```
I just elaborated in the comment.

robertnishihara commented:
> * `open_input_stream` itself **does NOT slow down seekable images**.

That's good to hear. Thanks for checking!
